src/server/middleware/blob/post_initiate_blob_upload.go

```diff
@@ -0,0 +1,65 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package blob
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/goharbor/harbor/src/api/blob"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/pkg/distribution"
+	"github.com/goharbor/harbor/src/server/middleware"
+)
+
+// PostInitiateBlobUploadMiddleware middleware to add blob to project after mount blob success
+func PostInitiateBlobUploadMiddleware() func(http.Handler) http.Handler {
+	skipper := middleware.NegativeSkipper(
+		middleware.MethodAndPathSkipper(http.MethodPost, distribution.InitiateBlobUploadRegexp),
+	)
+
+	return middleware.AfterResponse(func(w http.ResponseWriter, r *http.Request, statusCode int) error {
+		if statusCode != http.StatusCreated {
+			return nil
+		}
+
+		query := r.URL.Query()
+
+		mount := query.Get("mount")
+		if mount == "" {
+			return nil
+		}
+
+		logPrefix := fmt.Sprintf("[middleware][%s][blob]", r.URL.Path)
+
+		ctx := r.Context()
+		p, err := projectController.GetByName(ctx, distribution.ParseName(r.URL.Path))
+		if err != nil {
+			log.Errorf("%s: get project failed, error: %v", logPrefix, err)
+			return err
+		}
+
+		params := blob.EnsureParams{
+			MountDigest: mount,
+			ProjectID:   p.ProjectID,
+		}
+		if _, err := blobController.Ensure(ctx, params); err != nil {
```

Comment:
the conflict error is ignored?

Reply:
Yes, in dao the  insert ignore the conflict on the digest column.

Reply:
conflict error of insert will cause transaction failed, see more https://www.postgresql.org/message-id/002e01c04da9%24a8f95c20%2425efe6c1%40lasting.ro and https://stackoverflow.com/questions/41513125/can-postgres-silently-ignore-column-constraint-conflicts